### ISSUE TITLE
Multi-level Find for Maps with Comparer

### ISSUE THREAD

**Issue by StefanBertels** [louthy/language-ext#236]
The current readme has an example for finding a value in a nested map by calling outerMap.Find(outerKey, ..., innerKey). Seems this still exists --but only for Maps without comparer.

I have a `Map<OrdStringOrdinalIgnoreCase, string, Map<OrdStringOrdinalIgnoreCase, string, string>>` representing content of an INI file.

This is what I currently use to get a numeric value as Option<int>:
```
var value= ini.Find("Section").Bind(section=> section.Find("varname")).Bind(parseInt); 
```

This would be nice:
```
var value= ini.Find("Section", "varname").Bind(parseInt); 
```

**Comment by louthy** [comment]
Obiviously you can use LINQ to make it slightly prettier:
```c#
    var value = from section in ini.Find("Section")
                from varname in section.Find("varname")
                from result  in parseInt(varname)
                select result;
```
But yes, I will add the same extensions that the regular `Map` has.  Might be a few days though, I have quite a lot on at the moment.

**Comment by bender2k14** [comment]
The existing `Find` methods (that are extensions on nested `Map`s and take in all the keys at once) return an `Option<>` of the innermost key.  However, they are defined using a method with signature `R Find<R>(K key, Func<V, R> Some, Func<R> None)`.  I was surprised to see this.  I was expecting to see them defined by a method like `Option<V> Find(K key)`.

Why do these methods exist that combine the essential logic of `Find` with the logic of `Option<>.Match`?

It seems like there is an enormous amount of `Option<>.Match`-like code duplicated in order to implement these methods.  I prefer calling methods with fewer argurments, so I would prefer to call a `Find(key)` followed by `Match(some, none)` instead of `Find(key, some, none)`.  It also makes it awakard to try and define (say)
```
Option<D> Find<A, B, C, D>(this Map<A, Map<B, Map<C, D>>> self, A aKey, B bKey, C cKey)
```
in terms of
```
Option<C> Find<A, B, C>(this Map<A, Map<B, C>> self, A outerKey, B innerKey)
```
which is how I was expecting the existing `Find` (`Map` extension methods) would be defined.

**Comment by louthy** [comment]
@bender2k14 There's [overloads for both](https://github.com/louthy/language-ext/blob/master/LanguageExt.Core/DataTypes/Map/Map.Extensions.cs#L45).  So you can find by keys and get an `Option` or find by keys with an embedded `Some` `None` match.

The reason for the embedded match was mostly drying up a real world repetition I saw a lot of:
```c#
   var x = map.Find(key).Match(Some: x => ..., None: () => ...);
```
It was so common that I decided to create an overload to reduce the cognitive load:
```c#
   var x = map.Find(key, Some: x => ..., None: () => ...);
```
Perhaps calling it `FindMap` or `FindMatch` or something like that might be clearer.

**Comment by bender2k14** [comment]
It is interesting to hear which one has a smaller cognitive load for you.  It goes the other way around for me.  In fact, I am not that crazy about `.Match(some, none)` either.  I prefer `.Map(some).IfNone(none)` instead.

**Comment by StefanBertels** [comment]
my 2 pennies:

Being explict (e.g. FindMatch) has advantages:

1. It's easier to understand. Semantics defined by function name, not by signature.
2. improves type inference
3. makes learning functional programming easier because you can see that your using a composition (short cut) function.

**Comment by louthy** [comment]
Cognitive load was probably the wrong term.  'Less clutter' is probably more accurate.  There is precedence in the `Map` type for inline behaviour based on a  key's existence: `TryAdd`, `AddOrUpdate`, etc.  So, I think in this case it's a reasonable interface extension.  It isn't something I'd want to litter the code base with though.

> I am not that crazy about .Match(some, none)

There are efficiency benefits to `.Match(some, none)` over `.Map(some).IfNone(none)` which in some situations might be important.  It also looks a lot like the match statement of F#.  

Personally, I prefer to use `Match`, `IfNone`, etc. as little as possible and stay bound in the context of the monad.  It makes composition and propagation much easier.

There is also `opt.Some(some).None(none)` if you want a slightly more elegant fluent interface.

Although this library is clearly opinionated, there are a number of styles it supports to suit preferences of the developer.

* Function based  - `match(option, Some: some, None: none)`
* Fluent - `option.Some(some).None(none)`
* 'Pattern matching' - option.Match(Some: some, None: none)`

It's a policy I think works well, and even if you prefer one style, sometimes the others can be more elegant in certain situations (_Function based_ when used with LINQ for example)

> Being explict (e.g. FindMatch) has advantages:

In hindsight, I agree.  In reality it becomes a breaking change.  Although I guess I could make the `Find` extensions `Obsolete` and just add the newly named functions.

I'm midway through an enormous refactor for the async work - so I'll leave it for now and come back to it.  Not entirely sure `FindMatch` is the right name yet, so I'll have a think (open to suggestions).

**Comment by louthy** [comment]
@StefanBertels Do you fancy having a go at the implementation for this?  I'm pretty time poor at the moment, so a pull request would be lovely - rather than leaving this perpetually open :)

**Comment by StefanBertels** [comment]
Hi Paul,

I did some small experiments, current state can be seen here: https://github.com/StefanBertels/language-ext/commit/2974a7a5ff0d5c55d29080ed2bac298427f58dbf

There are multiple ways we can go and I'm not happy with any of them. Maybe I'm missing something.

Two side notes:
- I'd (still) favor having another name for Find function with built-in match (see discussion above).
- Maybe even multi level find could get another name, making multi-level access more visible.
Both things are just a matter of taste.

Now for the real thing:

It's probably easy to create multi level Find for each existing type itself, meaning we would have Multi-Level find for nested Maps with ORD on every level. But then we would have many types that would not get some nested Find function just because they don't have same properties. This does not only apply to Map<K,V> and Map<ORD,K,V> but also to HashMap<K,V> and HashMap<EQ,K,V>.

Would be quite useful to have multi level find for any combination of them.

Maybe one solution is to create some code generator producing every combination up to level n (e.g. n=5). I didn't write tt files before but I don't like this solution. But should work.

Another solution I tried is to have some new interface Searchable (Findable / TypeWithFind) that could enable any type with some Find method for multi level find. I fiddled with this in my [experiment commit](https://github.com/StefanBertels/language-ext/commit/2974a7a5ff0d5c55d29080ed2bac298427f58dbf).

Nice thing is, we only need to implement one extension function per supported level.
(Really) bad thing is that type inference doesn't work which makes it quite ugly (useless) in practise.

I fiddled a bit with making some covariant interface (see code), but probably I don't understand enough how things work here -- seems this doesn't help here.

Another thing that came to my mind is that it might be an acceptable solution to have some fluent interface for multi level find (again see code snippet in commit). But I didn't get something working here yet. My goal was to add some `ThenFind` function for every Option<Map<>> (and HashMap<...>) that would allow: `multimap.Find(key1).ThenFind(key2).ThenFind(key3)`

(Unsure whether ThenFind should be named Find, probably not because it combines Bind+Find.)

Any ideas for some solution that is usable in practise? Or should I try to create some code generator?
